Vamos a trabajar con datos de tiempos de juego para distintos juegos en base a un dataset scrapeado de "How Long To Beat"

In [ ]:
!wget https://raw.githubusercontent.com/KasumiL5x/hltb-scraper/master/all-games.csv
!wget https://raw.githubusercontent.com/KasumiL5x/hltb-scraper/master/all-completions.csv
!wget https://raw.githubusercontent.com/KasumiL5x/hltb-scraper/master/all-games-processed.csv

--2023-09-05 18:09:29--  https://raw.githubusercontent.com/KasumiL5x/hltb-scraper/master/all-games.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4831810 (4.6M) [text/plain]
Saving to: ‘all-games.csv’

all-games.csv       100%[===================>]   4.61M  --.-KB/s    in 0.05s   

2023-09-05 18:09:29 (87.1 MB/s) - ‘all-games.csv’ saved [4831810/4831810]

--2023-09-05 18:09:29--  https://raw.githubusercontent.com/KasumiL5x/hltb-scraper/master/all-completions.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 24835277 (24M) [

In [ ]:
import pandas as pd
import numpy as np

El primer dataframe que vamos a usar, es el de todos los juegos, que tiene para cada juego el tiempo promedio que toma completarlo entre otros datos.

In [ ]:
pd.read_csv("all-games-processed.csv")

id                        title main_story main_plus_extras  \
0      19962                          NaN        NaN              NaN   
1      38799                          NaN        NaN              NaN   
2      53256                   !4RC4N01D!        NaN              NaN   
3      53257  !4RC4N01D! 2: Retro Edition        NaN              NaN   
4      54173     !4RC4N01D! 3: Cold Space        NaN              NaN   
...      ...                          ...        ...              ...   
35917  52081                     ШП - ShP        NaN              NaN   
35918  38587            大海战 Navy Field IV        NaN              NaN   
35919  48480                 猫肉「Cat Meat」        NaN              NaN   
35920  65309                  音灵 INVAXION        NaN              NaN   
35921  63988                   ＜/reality＞        NaN              NaN   

            completionist          all_styles  coop  versus type  \
0                     NaN                 NaN   NaN     NaN  NaN   
1                     NaN                 NaN   NaN     NaN  NaN   
2      0.5666666666666667  0.5666666666666667   NaN     NaN  NaN   
3                       1                   1   NaN     NaN  NaN   
4                     1.5                 1.5   NaN     NaN  NaN   
...                   ...                 ...   ...     ...  ...   
35917                 1.5                 1.5   NaN     NaN  NaN   
35918                 NaN                 NaN   NaN     NaN  NaN   
35919                   4                   4   NaN     NaN  NaN   
35920                 NaN                 NaN   NaN     NaN  NaN   
35921                 NaN                 NaN   NaN     NaN  NaN   

                  developers              publishers platforms  \
0          Red Entertainment  Microsoft Game Studios       NaN   
1           Stormcloud Games                     NaN       NaN   
2                  armogames               armogames        PC   
3                        NaN                     NaN        PC   
4                  Armogames               Armogames        PC   
...                      ...                     ...       ...   
35917                Xitilon                 Xitilon        PC   
35918            SD EnterNET                     NaN        PC   
35919  2D Heroes, Ursa Games               2D Heroes        PC   
35920       Nanjing AQUATRAX                bilibili       NaN   
35921       Fancy Fish Games        Fancy Fish Games       NaN   

                                           genres         release_na  \
0                                      Simulation                NaN   
1      Third-Person, Action, Adventure, Roguelike    August 09, 2016   
2                                          Arcade   January 12, 2018   
3                                             NaN                NaN   
4                                          Arcade               2018   
...                                           ...                ...   
35917                                      Puzzle   October 31, 2017   
35918                                         NaN                NaN   
35919                                    Platform     April 21, 2017   
35920                                         NaN  December 19, 2018   
35921                                Visual Novel     April 18, 2017   

              release_eu      release_jp  
0                    NaN  April 24, 2003  
1        August 09, 2016             NaN  
2       January 12, 2018             NaN  
3                    NaN             NaN  
4                   2018             NaN  
...                  ...             ...  
35917   October 31, 2017             NaN  
35918                NaN   June 03, 2016  
35919     April 21, 2017             NaN  
35920  December 19, 2018             NaN  
35921     April 18, 2017             NaN  

[35922 rows x 16 columns]

Para simplificarnos la vida, vamos a dropear las filas que no tengan o nombre del juego o no tengan un valor en la columna de main_story (esto ultimo porque si)

In [ ]:
pd.read_csv("all-games-processed.csv").dropna(subset=["title","main_story"])

id                           title          main_story  \
15     54476                       #WarGames                   2   
16     45636                   #monstercakes                0.25   
17     40541                         $1 Ride  0.6166666666666667   
18        18                   'Splosion Man                   8   
19     38074             'n Verlore Verstand                   6   
...      ...                             ...                 ...   
35900  16464                           suteF                   2   
35906  49719                        unWorded                   2   
35911  56591             while True: learn()                 7.5   
35912  11419                             xXx                   3   
35915  13449  ¡Qué pasa, NENG! El Videojuego                   1   

      main_plus_extras completionist           all_styles  coop  versus type  \
15                   3           NaN                  2.5   NaN     NaN  NaN   
16                 NaN           0.4  0.36666666666666664   NaN     NaN  NaN   
17                 NaN           NaN   0.6166666666666667   NaN     NaN  NaN   
18                  10            18                    9   NaN     NaN  NaN   
19                 NaN           NaN                    6   NaN     NaN  NaN   
...                ...           ...                  ...   ...     ...  ...   
35900              NaN           NaN                    2   NaN     NaN  NaN   
35906                3           NaN                  2.5   NaN     NaN  NaN   
35911              8.5            11                    9   NaN     NaN  NaN   
35912              NaN           3.5                  3.5   NaN     NaN  NaN   
35915              NaN           1.5                  1.5   NaN     NaN  NaN   

                developers               publishers           platforms  \
15                     Eko                      Eko                  PC   
16            Paleno Games             Paleno Games                  PC   
17         BeHappy Studios    Back To Basics Gaming                  PC   
18     Twisted Pixel Games  Microsoft Games Studios  Xbox 360, Xbox One   
19         Skobbejak Games          Skobbejak Games                  PC   
...                    ...                      ...                 ...   
35900       Ted Lauterbach                      NaN                  PC   
35906         Bento Studio          Plug In Digital                 NaN   
35911             Luden.io                    Nival                  PC   
35912                  NaN               Activision    Game Boy Advance   
35915                  NaN                      NaN       PlayStation 2   

                           genres         release_na        release_eu  \
15            On-Rails, Adventure     March 14, 2018    March 14, 2018   
16                            NaN       May 05, 2017      May 05, 2017   
17                         Action      July 28, 2016               NaN   
18             Action, Platformer      July 22, 2009     July 22, 2009   
19     Virtual Reality, Adventure  February 01, 2016               NaN   
...                           ...                ...               ...   
35900            Platform, Puzzle                NaN               NaN   
35906                   Adventure   October 10, 2017  October 10, 2017   
35911     Educational, Simulation     March 28, 2018    March 28, 2018   
35912   Action, Action, Adventure               2002               NaN   
35915                         NaN                NaN               NaN   

      release_jp  
15           NaN  
16           NaN  
17           NaN  
18           NaN  
19           NaN  
...          ...  
35900        NaN  
35906        NaN  
35911        NaN  
35912        NaN  
35915        NaN  

[17324 rows x 16 columns]

Y como no me interesan todas las columnas, me quedo unicamente con la de id, el nombre del juego (title) y el tiempo para completar la historia principal (main_story)

In [ ]:
games = pd.read_csv("all-games-processed.csv").dropna(subset=["title","main_story"])[["id","title","main_story"]]
games

id                           title          main_story
15     54476                       #WarGames                   2
16     45636                   #monstercakes                0.25
17     40541                         $1 Ride  0.6166666666666667
18        18                   'Splosion Man                   8
19     38074             'n Verlore Verstand                   6
...      ...                             ...                 ...
35900  16464                           suteF                   2
35906  49719                        unWorded                   2
35911  56591             while True: learn()                 7.5
35912  11419                             xXx                   3
35915  13449  ¡Qué pasa, NENG! El Videojuego                   1

[17324 rows x 3 columns]

Tenemos un segundo csv con cada tiempo en particular que le tomo a cada persona completar un juego, en una plataforma especifica y una parte especifica (por ej, la historia principal)

In [ ]:
pd.read_csv("all-completions.csv")

id            type  platform      time
0          1      Main Story       NaN    3h 30m
1          1      Main Story        PC    7h 34m
2          1      Main Story        PC        9h
3          1      Main Story        PC   10h 08m
4          1      Main Story        PC   11h 01m
...      ...             ...       ...       ...
795678  9999  Completionists        PC  158h 23m
795679  9999  Completionists  Xbox 360      160h
795680  9999  Completionists        PC      170h
795681  9999  Completionists       NaN      180h
795682  9999  Completionists  Xbox 360      222h

[795683 rows x 4 columns]

Igual que antes, me voy a quedar con los tiempos para la historia principal

In [ ]:
df = pd.read_csv("all-completions.csv")
df = df[df.type == "Main Story"]
df

id        type platform     time
0          1  Main Story      NaN   3h 30m
1          1  Main Story       PC   7h 34m
2          1  Main Story       PC       9h
3          1  Main Story       PC  10h 08m
4          1  Main Story       PC  11h 01m
...      ...         ...      ...      ...
795610  9999  Main Story       PC      70h
795611  9999  Main Story       PC  72h 12m
795612  9999  Main Story       PC      75h
795613  9999  Main Story       PC      80h
795614  9999  Main Story       PC     100h

[384030 rows x 4 columns]

Notar que en este caso, el tiempo de juego esta representado como una cadena (bastante fea) a diferencia del otro DataFrame. Tengo que cambiar el tipo, voy a usar una funcion con regex y otras cosas, no le den mucha bola, en la clase de NLP les va a quedar mas claro.

In [ ]:
import re
import numpy as np

def parse_time_to_decimal_hours(time_str):
    match = re.match(r'(?:(\d+)h )?(\d+)m|(\d+)h', time_str)

    if match:
        hours = int(match.group(1) or 0)
        minutes = int(match.group(2) or 0)
        decimal_hours = hours + minutes / 60.0
        return decimal_hours
    else:
        return np.nan

df["time"] = df.time.map(parse_time_to_decimal_hours)

Hasta ahora no hice nada, solo procese 2 dataframes. Es medio aburrido, pero siempre hay que hacerlo antes de arrancar.

Bueno, ahora vamos a querer hacer es, buscar los juegos que mas tiempo alguien jugo y los que menos.

Primera cosa que deberia hacer es dropear los nan, si no tienen id, no me sirven, si no tienen tiempo, no tienen sentido. **OJO** No tiene sentido llenar los nans con 0 aca, que tengan un NaN me dice algo y no es que es un 0.

In [ ]:
df.dropna()

id        type platform       time
1          1  Main Story       PC   7.566667
2          1  Main Story       PC   0.000000
3          1  Main Story       PC  10.133333
4          1  Main Story       PC  11.016667
5          1  Main Story       PC  13.833333
...      ...         ...      ...        ...
795610  9999  Main Story       PC   0.000000
795611  9999  Main Story       PC  72.200000
795612  9999  Main Story       PC   0.000000
795613  9999  Main Story       PC   0.000000
795614  9999  Main Story       PC   0.000000

[378009 rows x 4 columns]

Para los mas jugados, podria ordenar todo y quedarme con los ultimos 5

In [ ]:
df.dropna().sort_values("time").tail(5)

id        type platform         time
85972   11720  Main Story       PC   841.316667
309949  26347  Main Story       PC  1024.183333
75012   11285  Main Story       PC  1083.366667
75014   11285  Main Story       PC  1327.533333
576300   5307  Main Story       PC  4512.433333

Funciona? si. Esta bien? no.

Tengo una alternativa para no tener que ordenar y es usar nlargest. Vemos que nos da el mismo resultado

In [ ]:
df.dropna().nlargest(5,"time")

id        type platform         time
576300   5307  Main Story       PC  4512.433333
75014   11285  Main Story       PC  1327.533333
75012   11285  Main Story       PC  1083.366667
309949  26347  Main Story       PC  1024.183333
85972   11720  Main Story       PC   841.316667

Es mucho muy importante esto. Hacer el ordenamiento es mas caro en tiempo que usar la alternatia de nlargest que hace un topK implementado con un heap.


Basicamente para encontrar los 5 mas grandes, generamos un heap de minimos (de tamaño 5), y recorremos los elementos metiendo uno en el heap y sacando el min. El orden entonces nos queda

* Agregar en un heap de tamaño K es O(log(k))
* Sacar el minimo en un heap de tamaño K es O(log(k))
* Voy a agregar y sacar N veces

El orden de buscar el top K usando un heap es

`O( n * 2 * log(k)) ==> O(nlogk)`

y como k es constante

`O(nlogk) ==> O(n)`

Ordenar, unicamente, es O(nlogn)

In [ ]:
largest = df.dropna().nlargest(5,"time")

Para los mas chicos tenemos la alternativa de usar nsmallest que funciona igual

In [ ]:
smallest = df.dropna().nsmallest(5,"time")

Con esto ahora puedo concatenarlos y tener un dataframe con ambos datos. Por simplicidad no los marco, porque no me importa para lo que sigue, pero se puede agregar un tag de a que grupo corresponde o hacer las cosas por separado directamente.

In [ ]:
concat = pd.concat([largest,smallest])
concat

id        type  platform         time
576300   5307  Main Story        PC  4512.433333
75014   11285  Main Story        PC  1327.533333
75012   11285  Main Story        PC  1083.366667
309949  26347  Main Story        PC  1024.183333
85972   11720  Main Story        PC   841.316667
2           1  Main Story        PC     0.000000
19       1000  Main Story  Xbox 360     0.000000
28      10001  Main Story       Wii     0.000000
32      10001  Main Story       Wii     0.000000
39      10001  Main Story       Wii     0.000000

### Parentesis

Que pasa si ahora trato de concatenar con axis = 1

In [ ]:
concat_axis_1 = pd.concat([largest,smallest],axis=1)
concat_axis_1

id        type platform         time       id        type  \
576300   5307.0  Main Story       PC  4512.433333      NaN         NaN   
75014   11285.0  Main Story       PC  1327.533333      NaN         NaN   
75012   11285.0  Main Story       PC  1083.366667      NaN         NaN   
309949  26347.0  Main Story       PC  1024.183333      NaN         NaN   
85972   11720.0  Main Story       PC   841.316667      NaN         NaN   
2           NaN         NaN      NaN          NaN      1.0  Main Story   
19          NaN         NaN      NaN          NaN   1000.0  Main Story   
28          NaN         NaN      NaN          NaN  10001.0  Main Story   
32          NaN         NaN      NaN          NaN  10001.0  Main Story   
39          NaN         NaN      NaN          NaN  10001.0  Main Story   

        platform  time  
576300       NaN   NaN  
75014        NaN   NaN  
75012        NaN   NaN  
309949       NaN   NaN  
85972        NaN   NaN  
2             PC   0.0  
19      Xbox 360   0.0  
28           Wii   0.0  
32           Wii   0.0  
39           Wii   0.0

Pasa lo mismo que con el indice de columnas, si hay un valor que coincide, agrega los datos ahi, sino, crea una columna. Notar que seria equivalente a concatenar las transpuestas

In [ ]:
pd.concat([largest.T,smallest.T])

576300       75014        75012        309949      85972   \
id               5307        11285        11285        26347       11720   
type       Main Story   Main Story   Main Story   Main Story  Main Story   
platform           PC           PC           PC           PC          PC   
time      4512.433333  1327.533333  1083.366667  1024.183333  841.316667   
id                NaN          NaN          NaN          NaN         NaN   
type              NaN          NaN          NaN          NaN         NaN   
platform          NaN          NaN          NaN          NaN         NaN   
time              NaN          NaN          NaN          NaN         NaN   

              2           19          28          32          39      
id               NaN         NaN         NaN         NaN         NaN  
type             NaN         NaN         NaN         NaN         NaN  
platform         NaN         NaN         NaN         NaN         NaN  
time             NaN         NaN         NaN         NaN         NaN  
id                 1        1000       10001       10001       10001  
type      Main Story  Main Story  Main Story  Main Story  Main Story  
platform          PC    Xbox 360         Wii         Wii         Wii  
time             0.0         0.0         0.0         0.0         0.0

In [ ]:
pd.concat([largest.T,smallest.T]).T

id        type platform         time     id        type  platform  \
576300   5307  Main Story       PC  4512.433333    NaN         NaN       NaN   
75014   11285  Main Story       PC  1327.533333    NaN         NaN       NaN   
75012   11285  Main Story       PC  1083.366667    NaN         NaN       NaN   
309949  26347  Main Story       PC  1024.183333    NaN         NaN       NaN   
85972   11720  Main Story       PC   841.316667    NaN         NaN       NaN   
2         NaN         NaN      NaN          NaN      1  Main Story        PC   
19        NaN         NaN      NaN          NaN   1000  Main Story  Xbox 360   
28        NaN         NaN      NaN          NaN  10001  Main Story       Wii   
32        NaN         NaN      NaN          NaN  10001  Main Story       Wii   
39        NaN         NaN      NaN          NaN  10001  Main Story       Wii   

       time  
576300  NaN  
75014   NaN  
75012   NaN  
309949  NaN  
85972   NaN  
2       0.0  
19      0.0  
28      0.0  
32      0.0  
39      0.0

Ya que estamos, sabian que pueden a fillna pasarle otra columna para que llene los valores faltantes con los de esta?

In [ ]:
concat_axis_1.columns = ["id","type","platform","time","A","B","C","D"]
concat_axis_1

id        type platform         time        A           B  \
576300   5307.0  Main Story       PC  4512.433333      NaN         NaN   
75014   11285.0  Main Story       PC  1327.533333      NaN         NaN   
75012   11285.0  Main Story       PC  1083.366667      NaN         NaN   
309949  26347.0  Main Story       PC  1024.183333      NaN         NaN   
85972   11720.0  Main Story       PC   841.316667      NaN         NaN   
2           NaN         NaN      NaN          NaN      1.0  Main Story   
19          NaN         NaN      NaN          NaN   1000.0  Main Story   
28          NaN         NaN      NaN          NaN  10001.0  Main Story   
32          NaN         NaN      NaN          NaN  10001.0  Main Story   
39          NaN         NaN      NaN          NaN  10001.0  Main Story   

               C    D  
576300       NaN  NaN  
75014        NaN  NaN  
75012        NaN  NaN  
309949       NaN  NaN  
85972        NaN  NaN  
2             PC  0.0  
19      Xbox 360  0.0  
28           Wii  0.0  
32           Wii  0.0  
39           Wii  0.0

In [ ]:
concat_axis_1.columns = ["id","type","platform","time","A","B","C","D"]
concat_axis_1.id.fillna(concat_axis_1.A)

576300     5307.0
75014     11285.0
75012     11285.0
309949    26347.0
85972     11720.0
2             1.0
19         1000.0
28        10001.0
32        10001.0
39        10001.0
Name: id, dtype: float64

 ### Fin del parentesis

 Vamos entonces a mergear los 2 dataframes para obtener los nombres

In [ ]:
pd.merge(concat,games,on="id")

id        type  platform  time                 title main_story
0      1  Main Story        PC   0.0  688(I) Hunter/Killer        9.5
1   1000  Main Story  Xbox 360   0.0           Bejeweled 2        1.5
2  10001  Main Story       Wii   0.0        The Last Story       21.5
3  10001  Main Story       Wii   0.0        The Last Story       21.5
4  10001  Main Story       Wii   0.0        The Last Story       21.5

momento, me faltan cosas

In [ ]:
pd.merge(concat,games,on="id",how="right")

id type platform  time                           title  \
0      54476  NaN      NaN   NaN                       #WarGames   
1      45636  NaN      NaN   NaN                   #monstercakes   
2      40541  NaN      NaN   NaN                         $1 Ride   
3         18  NaN      NaN   NaN                   'Splosion Man   
4      38074  NaN      NaN   NaN             'n Verlore Verstand   
...      ...  ...      ...   ...                             ...   
17321  16464  NaN      NaN   NaN                           suteF   
17322  49719  NaN      NaN   NaN                        unWorded   
17323  56591  NaN      NaN   NaN             while True: learn()   
17324  11419  NaN      NaN   NaN                             xXx   
17325  13449  NaN      NaN   NaN  ¡Qué pasa, NENG! El Videojuego   

               main_story  
0                       2  
1                    0.25  
2      0.6166666666666667  
3                       8  
4                       6  
...                   ...  
17321                   2  
17322                   2  
17323                 7.5  
17324                   3  
17325                   1  

[17326 rows x 6 columns]

Si al right le hacemos un dropna, terminamos en la misma situacion que antes (ergo, no hagan left o right para hacer dropna)

In [ ]:
pd.merge(concat,games,on="id",how="right").dropna()

id        type  platform  time                 title main_story
150        1  Main Story        PC   0.0  688(I) Hunter/Killer        9.5
1588    1000  Main Story  Xbox 360   0.0           Bejeweled 2        1.5
14866  10001  Main Story       Wii   0.0        The Last Story       21.5
14867  10001  Main Story       Wii   0.0        The Last Story       21.5
14868  10001  Main Story       Wii   0.0        The Last Story       21.5

In [ ]:
pd.merge(concat,games,on="id",how="left")

id        type  platform         time                 title main_story
0   5307  Main Story        PC  4512.433333                   NaN        NaN
1  11285  Main Story        PC  1327.533333                   NaN        NaN
2  11285  Main Story        PC  1083.366667                   NaN        NaN
3  26347  Main Story        PC  1024.183333                   NaN        NaN
4  11720  Main Story        PC   841.316667                   NaN        NaN
5      1  Main Story        PC     0.000000  688(I) Hunter/Killer        9.5
6   1000  Main Story  Xbox 360     0.000000           Bejeweled 2        1.5
7  10001  Main Story       Wii     0.000000        The Last Story       21.5
8  10001  Main Story       Wii     0.000000        The Last Story       21.5
9  10001  Main Story       Wii     0.000000        The Last Story       21.5

Fijense que los datos no estan bien. Por empezar, los tiempos no pueden ser 0, menos si el promedio de la historia sea 21.5; mismo para los maximos, por que no hay valores para esos titulos?